hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/HoodieCatalystExpressionUtils.scala

```diff
@@ -78,14 +80,52 @@ object HoodieCatalystExpressionUtils {
    * NOTE: Projection of the row from [[StructType]] A to [[StructType]] B is only possible, if
    * B is a subset of A
    */
-  def generateUnsafeProjection(from: StructType, to: StructType): UnsafeProjection = {
-    val attrs = from.toAttributes
-    val attrsMap = attrs.map(attr => (attr.name, attr)).toMap
-    val targetExprs = to.fields.map(f => attrsMap(f.name))
+  def generateUnsafeProjection(sourceStructType: StructType, targetStructType: StructType): UnsafeProjection = {
+    val resolver = SQLConf.get.resolver
+    val attrs = sourceStructType.toAttributes
+    val targetExprs = targetStructType.fields.map { targetField =>
+      val attrRef = attrs.find(attr => resolver(attr.name, targetField.name))
+        .getOrElse(throw new AnalysisException(s"Wasn't able to match target field `${targetField.name}` to any of the source attributes ($attrs)"))
+
+      genProjectingExpression(attrRef, targetField.dataType)
+    }
 
     GenerateUnsafeProjection.generate(targetExprs, attrs)
   }
 
+  private def genProjectingExpression(sourceExpr: Expression,
+                                      targetDataType: DataType): Expression = {
+    checkState(sourceExpr.resolved)
+
+    // TODO support array, map
+    (sourceExpr.dataType, targetDataType) match {
+      case (sdt, tdt) if sdt == tdt =>
+        sourceExpr
+
+      case (sourceType: StructType, targetType: StructType) =>
+        val fieldValueExprs = targetType.fields.map { tf =>
```

Comment:
> Looks like a subset of nested fields may be taken during the projection, e.g., if the source has a {a.b, a.c, a.d} and the target has a.b, we only keep a.b instead of the whole StructType a. Does this happen or the caller of this function always makes sure the targetStructType is properly constructed to preserve the root-level field instead of a subset of nested fields?

It does happen. It's actually the reason for this PR -- previously it was only handling nested field projections, but `NestedSchemaPruning` could produce schemas w/ _nested_ fields being pruned as well. Therefore, we need to make sure we handle this appropriately when reading log-files (by projecting records into the new schema)

Reply:
Got it

Reply:
Realized that this is actually not the right approach and the problem is elsewhere -- problem was in the following:

 - Therefore the problem here was that we're simply not reading projected records from the Parquet -- and the reason for that was that in case when non-whitelisted RecordPayload is used -- we will fallback to reading full record, but we still were allowing `NestedSchemaPruning` to be applied nevertheless